include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_for.h

```diff
@@ -192,8 +192,20 @@ __parallel_for(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPolicy&&
     {
         if (__count >= __large_submitter::__estimate_best_start_size(__exec, __brick))
         {
-            return __large_submitter{}(std::forward<_ExecutionPolicy>(__exec), __brick, __count,
-                                       std::forward<_Ranges>(__rngs)...);
+            // Passing begin() of each range is needed for the icpx compiler to vectorize. The indirection introduced
+            // by our all / guard views interfere with compiler vectorization. At this point, we have ensured that
+            // input is contiguous and can be operated on directly. The begin() function for these views will return a
+            // pointer which is passed to the kernel.
```

Comment:
This seems important for performance across all of oneDPL.  Should we explore this more fully (add an issue for now)?
Seems like its possible we need one of the following:
1) Improve our views or work with the compiler so they can be properly vectorized 
or
2) Adjust our kernel submitters to pass iterator (and size where necessary) instead of ranges by default

Reply:
Yes, I will open an issue for this as it needs further investigation.

In some of our tests, I am now seeing UR crashes with this change (I didn't see this when benchmarking), so I will look into that first to ensure the approach is correct.

Reply:
I made a slight change to this section of code and the comment. The workaround cannot work with buffers as we cannot pull a pointer from a buffer while on the host. Instead, I have added the workaround for types that are passed directly (e.g. USM) and buffers will take the else branch.

I'll open an issue once I harden out some more of the details. There is also a good chance this is a compiler optimization issue.

Reply:
After deeper investigation and experimentation, the `begin()` workaround has been completely removed. There
was a temporary compiler performance regression that has been completely resolved. The compiler is able to vectorize through our views without issue, and we do not need to implement any workaround.